function-app/adb-to-purview/src/Function.Domain/Helpers/OlProcessing/Synapse/OlSynapseMessageEnrichment.cs

```diff
@@ -0,0 +1,107 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Function.Domain.Models.OL;
+using Function.Domain.Providers;
+using Microsoft.Extensions.Logging;
+
+namespace Function.Domain.Helpers
+{
+    public class OlSynapseMessageEnrichment : IOlMessageEnrichment
+    {
+        private ILogger _log;
+        private const string PLAN_CLASS = "org.apache.spark.sql.delta.commands.MergeIntoCommand";
+        private const string SOURCE_CLASS = "org.apache.spark.sql.execution.datasources.LogicalRelation";
+        private ISynapseClientProvider _synapseClientProvider;
+
+        public OlSynapseMessageEnrichment(ILoggerFactory loggerFactory, ISynapseClientProvider synapseClientProvider)
+        {
+            _log = loggerFactory.CreateLogger<OlSynapseMessageEnrichment>();
+            _synapseClientProvider = synapseClientProvider;
+        }
+
+        public async Task<Event?> EnrichmentEventAsync(Event olEvent, string workspaceName)
+        {
+            try
+            {
+                // Capture Name
+                var sourceNames = CaptureName(olEvent, "source");
+                var targetNames = CaptureName(olEvent, "target");
+
+                // Capture Namespace
+                List<Inputs> inputs = await CaptureNameSpaceAsync<Inputs>(sourceNames, workspaceName);
+                List<Outputs> outputs = await CaptureNameSpaceAsync<Outputs>(targetNames, workspaceName);
+
+                // Merge and create a distinct collection based on Name and NameSpace
+                if (inputs.Count > 0)
+                {
+                    olEvent.Inputs = MergeAndDistinct(olEvent.Inputs, inputs, x => x.Name, x => x.NameSpace);
+                }
+
+                if (outputs.Count > 0)
+                {
+                    olEvent.Outputs = MergeAndDistinct(olEvent.Outputs, outputs, x => x.Name, x => x.NameSpace);
+                }
+            }
+            catch (Exception ex)
+            {
+                _log.LogError(ex, "OlSynapseMessageEnrichment-EnrichmentEventAsync: ErrorMessage {ErrorMessage} ", ex.Message);
+                //throw;  // TODO Mani check if it breaks anywhere it will not continue
+            }
+            return olEvent;  // TODO Mani check Gaceful return
+        }
+
+        private async Task<List<T>> CaptureNameSpaceAsync<T>(HashSet<string> names, string workspaceName)
+        {
+            List<T> result = new List<T>();
+            // Define a delegate for the factory method
+            Func<string, string, T> factoryMethod = (Func<string, string, T>)Delegate.CreateDelegate(typeof(Func<string, string, T>), typeof(T).GetMethod("CreateInstance"));
```

Comment:
Let's avoid reflection if possible. You can use generics here and use a constraint on T and use the inteface and new T() syntax.

```
private T CreateInputsOutputs<T>(string name, string namespaceValue) where T : IInputsOutputs, new()
        {
            return new T() { Name = name, NameSpace = namespaceValue};
        }
```

Reply:
Done